query_execution.go

```diff
@@ -0,0 +1,889 @@
+				return nil, err
+			}
+		} else {
+			if field.Name != "__typename" {
+				return nil, fmt.Errorf("buildTypenameResponseMap: expected __typename")
+			}
+			result[field.Alias] = parentTypeName
+		}
+	}
+	return result, nil
+}
+
+func fragmentImplementsAbstractType(schema *ast.Schema, objectType, interfaceType string) bool {
+	for _, def := range schema.Implements[objectType] {
+		if def.Name == interfaceType {
+			return true
+		}
+	}
+	return false
+}
+
+func extractAndDedupeBoundaryIDs(data interface{}, insertionPoint []string) ([]string, error) {
+	boundaryIDs, err := extractBoundaryIDs(data, insertionPoint)
+	if err != nil {
+		return nil, err
+	}
+	dedupeMap := make(map[string]struct{}, len(boundaryIDs))
+	for _, boundaryID := range boundaryIDs {
+		dedupeMap[boundaryID] = struct{}{}
+	}
+
+	deduped := make([]string, 0, len(boundaryIDs))
+	for id := range dedupeMap {
+		deduped = append(deduped, id)
+	}
+
+	return sort.StringSlice(deduped), nil
+}
+
+func extractBoundaryIDs(data interface{}, insertionPoint []string) ([]string, error) {
+	ptr := data
+	if ptr == nil {
+		return nil, nil
+	}
+	if len(insertionPoint) == 0 {
+		switch ptr := ptr.(type) {
+		case map[string]interface{}:
+			id, err := boundaryIDFromMap(ptr)
+			return []string{id}, err
+		case []interface{}:
+			result := []string{}
+			for _, innerPtr := range ptr {
+				ids, err := extractBoundaryIDs(innerPtr, insertionPoint)
+				if err != nil {
+					return nil, err
+				}
+				result = append(result, ids...)
+			}
+			return result, nil
+		default:
+			return nil, fmt.Errorf("extractBoundaryIDs: unexpected type: %T", ptr)
+		}
+	}
+	switch ptr := ptr.(type) {
+	case map[string]interface{}:
+		return extractBoundaryIDs(ptr[insertionPoint[0]], insertionPoint[1:])
+	case []interface{}:
+		result := []string{}
+		for _, innerPtr := range ptr {
+			ids, err := extractBoundaryIDs(innerPtr, insertionPoint)
+			if err != nil {
+				return nil, err
+			}
+			result = append(result, ids...)
+		}
+		return result, nil
+	default:
+		return nil, fmt.Errorf("extractBoundaryIDs: unexpected type: %T", ptr)
+	}
+}
+
+func buildBoundaryQueryDocuments(ctx context.Context, schema *ast.Schema, step *QueryPlanStep, ids []string, parentTypeBoundaryField BoundaryField, batchSize int) ([]string, error) {
+	selectionSetQL := formatSelectionSetSingleLine(ctx, schema, step.SelectionSet)
+	if parentTypeBoundaryField.Array {
+		qids := []string{}
+		for _, id := range ids {
+			qids = append(qids, fmt.Sprintf("%q", id))
+		}
+		idsQL := fmt.Sprintf("[%s]", strings.Join(qids, ", "))
+		return []string{fmt.Sprintf(`{ _result: %s(ids: %s) %s }`, parentTypeBoundaryField.Field, idsQL, selectionSetQL)}, nil
+	}
+
+	var (
+		documents      []string
+		selectionIndex int
+	)
+	for _, batch := range batchBy(ids, batchSize) {
+		var selections []string
+		for _, id := range batch {
+			selection := fmt.Sprintf("%s: %s(id: %q) %s", nodeAlias(selectionIndex), parentTypeBoundaryField.Field, id, selectionSetQL)
```

Comment:
@lucianjon Should we retire the use of `nodeAlias` since we want to deprecate `node` boundary fields?

Reply:
Will drop the function as the name is no longer relevant but I think it makes sense to keep the behaviour of using `_{id}` as the key when doing non-array boundary field lookups. These are still the `@boundary` directive based lookups rather than the node ones.